src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -773,3 +773,40 @@ GetForeignKeyOids(Oid relationId, int flags)
 
 	return foreignKeyOids;
 }
+
+
+/*
+ * SetSkipValidationIfForeignKeyConstraint takes a Node object and if it's an
+ * AlterTableStmt, then sets skip_validation fields of
+ *  - "ALTER TABLE ADD FOREIGN KEY CONSTRAINT" and
+ *  - "ALTER TABLE ADD COLUMN %s REFERENCES"
+ * subcommands to skip their initial validation.
+ */
+void
+SkipValidationIfAddForeignKeyCommand(Node *stmt)
+{
+	if (!IsA(stmt, AlterTableStmt))
+	{
+		return;
+	}
+
+	AlterTableStmt *alterTableStatement = (AlterTableStmt *) stmt;
+
+	List *alterTableCommandList = alterTableStatement->cmds;
+
+	AlterTableCmd *alterTableCommand = NULL;
+	foreach_ptr(alterTableCommand, alterTableCommandList)
+	{
+		AlterTableType alterTableCommandType = alterTableCommand->subtype;
+
+		if (alterTableCommandType == AT_AddConstraint || alterTableCommandType ==
+			AT_AddColumn)
+		{
```

Comment:
Well, in the existing  code when it is `AT_AddColumn`, we do:
```
ColumnDef *columnDefinition = (ColumnDef *) command->def;
```

Now, we're casting it to 

```
Constraint *constraint = (Constraint *) alterTableCommand->def;
```

Do we trigger this case with a test?

Reply:
opps, it doesn't crash I guess but it should/might. I will look into that a bit closer

Reply:
Would you mind sharing an example table/commands to trigger `AT_AddColumn `? Thanks

Reply:
`ALTER TABLE table_x ADD COLUMN col_x int REFERENCES table_y(col_y);` can be an example.

The reason we are not hitting to any crash in this pr is that  citus does not allow defining fkeys in this way. But I need to fix it for sure :)

Reply:
```sql
postgres=# ALTER TABLE table_x ADD COLUMN col_x int REFERENCES table_y(col_y);
ERROR:  cannot execute ADD COLUMN command with PRIMARY KEY, UNIQUE, FOREIGN and CHECK constraints
DETAIL:  Adding a column with a constraint in one command is not supported because all constraints in Citus must have explicit names
HINT:  You can issue each command separately such as ALTER TABLE table_x ADD COLUMN col_x data_type; ALTER TABLE table_x ADD CONSTRAINT constraint_name FOREIGN KEY (col_x) REFERENCES table_y(col_y);
```

Reply:
Are we sure? The code-cov report tells  that we're triggering this line: https://codecov.io/gh/citusdata/citus/src/ce1c6a5f4e151440a5d4f1ac91f3e0b2ac5e3660/src/backend/distributed/commands/table.c#L398

Can you maybe add a WARNING or such to that line on master and run check-full (probably pushing to github and waiting CI is simpler) to see what triggers it?

Reply:
In https://github.com/citusdata/citus/tree/test-add-fkey-by-col, I put an `ereport` call with `ERROR` and observed that this line is executed actually, you are right.

As far as I see, before giving the error I shared in https://github.com/citusdata/citus/pull/3872#discussion_r435919359, the lines I made changes in this pr are anyway executed, but we somehow didn't get segfaults surprisingly.

Now, I think I fixed the possible crash that this pr might cause.